README.MD
# Description

This repo contains Python scripts to manage the actions I use most often with regards to AWS Lambda. Specifically, the scripts automate the following actions:

* Update a Lambda function's source code, env variables, and layers for the `$LATEST` alias.
* Publish a Lambda function to an alias or update alias's version if it already exists.
* Publish or update version of a Lambda layer.

Although these three actions are a far cry from full automation of working with Lambda, in my use case, they cover more than 90% of the interaction I have with Lambda. The usage and caveats of each script will be described below.

**_!The scripts are developed for Python-based AWS Lambda function. Lambda function written in other languages has NOT been tested!_**


# Recommended Project Layout

All the scripts are inside `scripts` folder. To use them without surprise, a project is highly recommended to be structured as follows:

```
├── lambda_functions
│   └── sample_lambda
│       ├── config.dev.json
│       ├── config.foo.json
│       ├── config.prod.json
│       └── lambda_function.py
└── scripts
    ├── __init__.py
    ├── config.py
    ├── publish_layer.py
    ├── publish_to_alias.py
    └── update_lambda.py
```

* `scripts/`: This folder holds all the scripts that handle the three AWS Lambda actions. Except from `boto3`, the scripts do NOT use any external packages.
* `lambda_functions`: This folder holds all the subfolders, each corresponding to a lambda function.
* `lambda_functions/sample_lambda`: This is an example folder for a lambda function of **the same name**. All Python code for the lambda function must live here.
* `lambda_functions/sample_lambda/config.dev.json`: This is the configuration for the development, or `$LATEST` alias, for the lambda function. It contains information for the env variables and a list of layer names. Example:
    ```json
    {
        "vars": {"env": "dev"},
        "layers": ["requests"]
    }
    ```
    The example shows that the dev version of the lambda function requires one env variable `env=dev`, and one already published layer called `requests`.
* `lambda_functions/sample_lambda/config.prod.json`: This, along with `config.foo.json`, are the configuration for each lambda alias. They only contain information about the env variables. The file layout is the same as shown above, except that there is no `"layers"` key.

# Usage

The usage described below assumes that the project layout from above is adopted.

## Update an EXISTING lambda function
The scripts **do NOT handle** the case of creating a new lambda. Thus, the initial creation of the lambda function, IAM role attachment, trigger and/or destination must be configured manually. This is the biggest caveate of these scripts.

Follow these steps to update an existing lambda function

1. Edit and test the function source code locally.
2. Create and edit the configuration of the dev (i.e. `$LATEST`) version of the function in `config.dev.json` file.
3. Use `scripts/update_lambda.py` to update the lambda function. Below is the help message of the script. See [example](https://github.com/FanchenBao/aws_lambda_scripts#example) for sample usage.
    ```
    ╰─$ python3 scripts/update_lambda.py -h
    usage: update_lambda.py [-h] --func_folder_path FUNC_FOLDER_PATH
                        --func_name FUNC_NAME --description DESCRIPTION

    Update lambda code and configuration for AWS Lambda Usage: env=dev
    python3 update_lambda --func_name [NAME] --func_folder_path [PATH]
    --description [DESC].

    optional arguments:
    -h, --help              show this help message and exit
    --func_folder_path FUNC_FOLDER_PATH
                            REQUIRED. The path to the folder containing one
                            or more Python files that contribute to the
                            lambda function
    --func_name FUNC_NAME
                            REQUIRED. The name of the lambda function as
                            shown on AWS.
    --description DESCRIPTION
                            REQUIRED. The description of the lambda function.
    ```

### Caveat
As mentioned already, the lambda function must have already been created for the update script to work properly.


## Publish a lambda function to an alias
Except from the word `dev`, there is no restriction on alias name. The reason `dev` is forbidden for alias name is to ensure that the function for dev is always tied to `$LATEST` version.

Follow these steps to publish to alias

1. Create and edit the configuration for the alias in `config.[alias].json` file.
2. Use `scripts/publish_to_alias.py` to publish the alias. Below is the help message of the script. See [example](https://github.com/FanchenBao/aws_lambda_scripts#example) for sample usage.
    ```
    ╰─$ python3 scripts/publish_to_alias.py -h
    usage: publish_to_alias.py [-h] --func_folder_path FUNC_FOLDER_PATH
                            --func_name FUNC_NAME --alias ALIAS
                            --description DESCRIPTION

    Publish a new version of the well-tested lambda function, and assign an
    alias to it. When publishing a new version, the only thing that needs to
    be updated is the environmental variable. Usage: python3 publish_to_alias
    --func_folder_path [path] --func_name [lambda name] --alias [alias]
    --description [description of the alias].

    optional arguments:
    -h, --help            show this help message and exit
    --func_folder_path FUNC_FOLDER_PATH
                            REQUIRED. The path to the folder containing
                            configs for lambda function of different alias
    --func_name FUNC_NAME
                            REQUIRED. The name of the lambda function as
                            shown on AWS.
    --alias ALIAS           REQUIRED. Name of the alias.
    --description DESCRIPTION
                            REQUIRED. The description of the lambda function
                            version used forthe alias.
    ```

## Publish a layer
Layer is a way for different lambda functions to share the same resources. Typically, one uploads a third-party package not offered by AWS Lambda's Python runtime as a layer, such that more than one lambda functions can import and use that package and there is no need to include that package in the lambda source code.

For more information about layers, check out its [documentation](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html).

Use `scripts/publish_layer.py` to publish a layer. Below is the help message of the scripts. See [example](https://github.com/FanchenBao/aws_lambda_scripts#example) for sample usage.

    ```
    ╰─$ python3 scripts/publish_layer.py -h
    usage: publish_layer.py [-h] --package_name PACKAGE_NAME --description
                            DESCRIPTION

    Publish a layer to AWS Lambda. A layer is a package that is shared and
    can be imported by all Lambda under the same account If the layer name
    never exists before, create a new one. If it already exists, update it to
    a new version. Note that layer name must be the same as package
    nameUsage: python3 publish_layer --package_name [PACKAGE
    NAME]--description [DESC].

    optional arguments:
    -h, --help              show this help message and exit
    --package_name PACKAGE_NAME
                            REQUIRED. Name of the package to be uploaded as a
                            layer.
    --description DESCRIPTION
                            REQUIRED. Description of the latest version of
                            the layer.
    ```

### Caveat
The method for publishing layer borrows from [this post](https://aws.amazon.com/premiumsupport/knowledge-center/lambda-import-module-error-python/). However, there is one major difference: in the post, the package is installed on AWS EC2. This is important because not all Python packages work across all platforms. For instance, since I am on macOS, when I use `scripts/publish_layer.py`, the published layer contains the package specifically installed for macOS. AWS Lambda function runs on Amazon Linux. If the Python package for macOS is compatible with Amazon Linux (e.g. `requests`), then our published layer works. However, if the Python package for macOS is not compatible with Amazon Linux (e.g. `scikit-learn`), one must install and publish the layer from an EC2 instance.

Overall, this shouldn't be too big an issue, because the amount of times one need to publish layer is much lower than update lambda function source code or publish to an alias.


# Example

This repo contains a sample lambda function to demonstrate how the scripts work. Follow the steps below to see for yourself.

1. Clone the repo
    ```bash
    git clone https://github.com/FanchenBao/aws_lambda_scripts.git
    ```
2. Create a virtualenv for `aws_lambda_scripts`.
3. Install dependencies
    ```bash
    pip install -r requirements.txt
    ```
4. Configure credentials for `boto3`. I recommend setting environment variables to specify which credentials to use ([documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#environment-variables)). For example, I export `AWS_PROFILE=default` to the shell to ensure that all lambda actions happen to my default AWS account.
5. Create a lambda function called `sample_lambda` on AWS. Assign it basic Lambda function execution role.
6. Publish a layer for the package `requests`. This external package will be used by the lambda function
    ```bash
    python3 scripts/publish_layer.py \
        --package_name requests \
        --description "A layer to use requests"
    ```
7. Update lambda function (alternatively, use `make update_sample_lambda`)
    ```bash
    env=dev python3 scripts/update_lambda.py \
        --func_folder_path lambda_functions/sample_lambda \
        --func_name sample_lambda \
        --description "A lambda function for demo purpose"
    ```
    After update is successful, go to AWS console, test the function, and examine the output.
8. Publish the lambda function to `prod` alias (alternatively, use `make publish_sample_lambda_to_prod`)
    ```bash
    python3 scripts/publish_to_alias.py \
		--func_folder_path lambda_functions/sample_lambda \
		--func_name sample_lambda \
		--alias prod \
		--description "Alias for production"
    ```
    After publication is successful, go to AWS console, inspect the `prod` alias, test the alias, and examine the output. Compared to the output from the `$LATEST` version, one shall see the difference in the env variable value.
9. (Optional) Publish the lambda function to another alias called `foo` (alternatively, use `make publish_sample_lambda_to_foo`)
    ```bash
    python3 scripts/publish_to_alias.py \
		--func_folder_path lambda_functions/sample_lambda \
		--func_name sample_lambda \
		--alias foo \
		--description "Just some random alias"
    ```
    Inspect and test the `foo` alias the same way as before.



CHANGELOG.MD
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [v0.0.1] - 2021-11-10
### Added
- The entire project. This is the first version.


scripts/publish_layer.py
# coding: utf-8

import logging
import shlex
import subprocess
import sys
from argparse import ArgumentParser
from pathlib import Path
from shutil import make_archive, rmtree

from config import CLIENT, LAYER_TEMP, LAYER_ZIP, PYTHON_VERSION, ROOT_DIR

# set up logger
logger = logging.getLogger()
console_handler = logging.StreamHandler()
formatter = logging.Formatter(
    '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
)
console_handler.setFormatter(formatter)
logger.addHandler(console_handler)
logger.setLevel(logging.INFO)


def clean_up() -> None:
    """Remove the temp folder and function zip file."""
    rmtree(LAYER_TEMP)
    Path.unlink(ROOT_DIR.joinpath(LAYER_ZIP))


def get_argument_parser() -> ArgumentParser:  # pragma no cover
    """Set up a parser to parse command line arguments.

    :return: A fresh, unused, ArgumentParser.
    """
    parser = ArgumentParser(
        description=(
            'Publish a layer to AWS Lambda. A layer is a package that is '
            'shared and can be imported by all Lambda under the same account '
            'If the layer name never exists before, create a new one. If it '
            'already exists, update it to a new version. Note that layer name '
            'must be the same as package name'
            'Usage: python3 publish_layer --package_name [PACKAGE NAME]'
            '--description [DESC].'
        ),
    )
    parser.add_argument(
        '--package_name',
        dest='package_name',
        type=str,
        required=True,
        help=(
            'REQUIRED. Name of the package to be uploaded as a layer.'
        ),
    )
    parser.add_argument(
        '--description',
        dest='description',
        type=str,
        required=True,
        help=(
            'REQUIRED. Description of the latest version of the layer.'
        ),
    )
    return parser


def exception_handler(error_msg: str, err: Exception) -> None:
    """Log exception and error message, and clean up the procedure.

    :param error_msg: The error message to log.
    :param err: The exception itself.
    """
    logger.exception(err)
    logger.error(f'\033[30;41m{error_msg}\033[0m')
    clean_up()
    sys.exit(1)


def prep_files(package_name: str) -> None:
    """Prepare a folder to hold all the files/folders for the package.

    Follow the instruction from here:
    https://aws.amazon.com/premiumsupport/knowledge-center/lambda-import-module-error-python/

    However, one major difference is that we are doing this on our local
    machine, which might or might not be incompatible with Amazon Linux
    machines.

    :param package_name: Name of the package to be zipped up for upload as
        layer.
    """
    # prepare temp folder
    if Path.exists(LAYER_TEMP):  # clean up any previous upload
        rmtree(LAYER_TEMP)
    Path.mkdir(LAYER_TEMP, parents=True)
    subprocess.run(
        shlex.split(f'pip install -t {LAYER_TEMP} {package_name}'),
    )
    # create zip file
    make_archive(
        LAYER_ZIP.split('.')[0],
        'zip',
        root_dir=ROOT_DIR,
        base_dir=LAYER_TEMP.parts[-1],
    )


def publish_layer(layer_name: str, desc: str) -> None:
    """Upload lambda code base (as a zip file) to its associated AWS Lambda.

    :param layer_name: Name of the layer.
    :param desc: Description of the layer.
    :raises RuntimeError: Either update layer fails
    """
    with open(LAYER_ZIP, 'rb') as f_obj:  # upload
        zip_bytes = f_obj.read()
    try:
        resp = CLIENT.publish_layer_version(
            LayerName=layer_name,
            Description=desc,
            Content={'ZipFile': zip_bytes},
            CompatibleRuntimes=[PYTHON_VERSION],
        )
    except Exception as err:
        raise RuntimeError('Publish layer FAILED.') from err
    logger.info(resp)


if __name__ == '__main__':
    # Parse command line arguments
    parser: ArgumentParser = get_argument_parser()
    args = parser.parse_args()

    logger.info(
        f'\033[30;43mPublishing {args.package_name} layer...\033[0m',
    )
    try:
        prep_files(args.package_name)
    except Exception as err2:
        exception_handler('Failed to prepare files', err2)

    try:
        publish_layer(args.package_name, args.description)
    except Exception as err3:
        exception_handler('Failed to publish layer', err3)

    clean_up()
    logger.info(
        f'\033[30;42mPublish {args.package_name} layer SUCCESS!\033[0m',
    )


scripts/update_lambda.py
# coding: utf-8

import json
import logging
import os
import sys
from argparse import ArgumentParser
from pathlib import Path
from shutil import copy, make_archive, rmtree
from typing import Dict, List, Union, cast

from config import (
    CLIENT,
    FUNCTION_TEMP,
    FUNCTION_ZIP,
    PYTHON_VERSION,
    ROOT_DIR,
)

# set up logger
logger = logging.getLogger()
console_handler = logging.StreamHandler()
formatter = logging.Formatter(
    '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
)
console_handler.setFormatter(formatter)
logger.addHandler(console_handler)
logger.setLevel(logging.INFO)


def clean_up() -> None:
    """Remove the temp folder and function zip file."""
    rmtree(FUNCTION_TEMP)
    Path.unlink(ROOT_DIR.joinpath(FUNCTION_ZIP))


def get_argument_parser() -> ArgumentParser:  # pragma no cover
    """Set up a parser to parse command line arguments.

    :return: A fresh, unused, ArgumentParser.
    """
    parser = ArgumentParser(
        description=(
            'Update lambda code and configuration for AWS Lambda '
            'Usage: env=dev python3 '
            'update_lambda --func_name [NAME] --func_folder_path [PATH] '
            '--description [DESC].'
        ),
    )
    parser.add_argument(
        '--func_folder_path',
        dest='func_folder_path',
        type=str,
        required=True,
        help=(
            'REQUIRED. The path to the folder containing one or more Python '
            'files that contribute to the lambda function'
        ),
    )
    parser.add_argument(
        '--func_name',
        dest='func_name',
        type=str,
        required=True,
        help=(
            'REQUIRED. The name of the lambda function as shown on AWS.'
        ),
    )
    parser.add_argument(
        '--description',
        dest='description',
        type=str,
        required=True,
        help=(
            'REQUIRED. The description of the lambda function.'
        ),
    )
    return parser


def exception_handler(error_msg: str, err: Exception) -> None:
    """Log exception and error message, and clean up the procedure.

    :param error_msg: The error message to log.
    :param err: The exception itself.
    """
    logger.exception(err)
    logger.error(f'\033[30;41m{error_msg}\033[0m')
    clean_up()
    sys.exit(1)


def prep_files(func_folder_path: str) -> None:
    """Prepare a temp folder to hold all the files to be uploaded to Lambda.

    :param func_folder_path: Path to the folder that holds all the Python files
        for the Lambda function.
    """
    # prepare temp folder
    if Path.exists(FUNCTION_TEMP):  # clean up any previous upload
        rmtree(FUNCTION_TEMP)
    Path.mkdir(FUNCTION_TEMP, parents=True)
    # copy files to temp
    for lambda_file in ROOT_DIR.joinpath(f'{func_folder_path}').glob('*.py'):
        copy(str(lambda_file), str(FUNCTION_TEMP))
    # create zip file
    make_archive(FUNCTION_ZIP.split('.')[0], 'zip', str(FUNCTION_TEMP))


def get_layer_arn(layers: List) -> List[str]:
    """Get the arns for each layer specified by the config JSON file.

    :param layers: Name of the layers specified in the config JSON file.
    :type layers: List
    :raises RuntimeError: when getting the list of layers fails
    :return: A list of layer arns.
    :rtype: List[str]
    """
    if not layers:
        return []
    try:
        resp = CLIENT.list_layers(CompatibleRuntime=PYTHON_VERSION)
    except Exception as err:
        raise RuntimeError('Unable to obtain ARNs for layers') from err
    res = []
    for ly in resp['Layers']:
        if ly['LayerName'] in layers:
            res.append(ly['LatestMatchingVersion']['LayerVersionArn'])
    return res


def update(
    func_name: str,
    desc: str,
    func_config: Dict[str, Union[Dict, List]],
) -> None:
    """Upload lambda code base (as a zip file) to its associated AWS Lambda.

    :param func_name: Name of the Lambda function as shown on AWS Lambda.
    :param desc: Description of the Lambda function.
    :param func_config: Configs associated with the Lambda function. It can
        have the following shape. {'vars': {key: value}, 'layers': [layers]}.
        The value for 'vars' is a Dict describing the key-value pair used by
        the Lambda function as environment variables. The value 'layers' is a
        list of layers used by the Lambda function. If no environment variable
        or no layer exists, func_config can be passed as an empty dict.
    :raises RuntimeError: Either update function code fails, or update function
        config fais.
    """
    with open(FUNCTION_ZIP, 'rb') as f_obj:  # upload
        zip_bytes = f_obj.read()
        try:
            resp1 = CLIENT.update_function_code(
                FunctionName=func_name,
                ZipFile=zip_bytes,
            )
        except Exception as err1:
            raise RuntimeError('Update lambda function code FAILED.') from err1
        logger.info(resp1)

    try:  # config
        resp2 = CLIENT.update_function_configuration(
            FunctionName=func_name,
            Description=desc,
            Environment={
                'Variables': func_config.get('vars', {}),
            },
            Layers=get_layer_arn(cast(List, func_config.get('layers', []))),
        )
    except Exception as err2:
        raise RuntimeError(
            'Update lambda function configuration FAILED.',
        ) from err2
    logger.info(resp2)


if __name__ == '__main__':
    # Parse command line arguments
    parser: ArgumentParser = get_argument_parser()
    args = parser.parse_args()

    logger.info(f'\033[30;43mUpdating code in {args.func_name}...\033[0m')

    if 'env' not in os.environ or os.environ['env'] != 'dev':
        logger.error('\033[30;41mMust use this script when "env=dev"!\033[0m')
        sys.exit(1)
    try:
        prep_files(args.func_folder_path)
    except Exception as err_prep:
        exception_handler('Failed to prepare zip file for function', err_prep)

    config_file = ROOT_DIR.joinpath(
        f'{args.func_folder_path}/config.dev.json',
    )
    try:
        with open(config_file, 'r') as f_obj:
            func_config = json.load(f_obj)
    except FileNotFoundError as err_config:
        exception_handler(f'Must provide {config_file} file', err_config)

    try:
        update(args.func_name, args.description, func_config)
    except Exception as err_update:
        exception_handler('Failed to update function', err_update)
    clean_up()
    logger.info(f'\033[30;42mUpdate {args.func_name} SUCCESS!\033[0m')


scripts/publish_to_alias.py
# coding: utf-8

import json
import logging
import sys
from argparse import ArgumentParser
from pathlib import Path
from typing import Dict

from config import CLIENT, ROOT_DIR

# set up logger
logger = logging.getLogger()
console_handler = logging.StreamHandler()
formatter = logging.Formatter(
    '%(asctime)s - %(name)s - %(levelname)s - %(message)s',
)
console_handler.setFormatter(formatter)
logger.addHandler(console_handler)
logger.setLevel(logging.INFO)


def get_argument_parser() -> ArgumentParser:  # pragma no cover
    """Set up a parser to parse command line arguments.

    :return: A fresh, unused, ArgumentParser.
    """
    parser = ArgumentParser(
        description=(
            'Publish a new version of the well-tested lambda function, and '
            'assign an alias to it. When publishing '
            'a new version, the only thing that needs to be updated is the '
            'environmental variable. '
            'Usage: python3 '
            'publish_to_alias --func_folder_path [path] '
            '--func_name [lambda name] '
            '--alias [alias] '
            '--description [description of the alias].'
        ),
    )
    parser.add_argument(
        '--func_folder_path',
        dest='func_folder_path',
        type=str,
        required=True,
        help=(
            'REQUIRED. The path to the folder containing configs for lambda '
            'function of different alias'
        ),
    )
    parser.add_argument(
        '--func_name',
        dest='func_name',
        type=str,
        required=True,
        help=(
            'REQUIRED. The name of the lambda function as shown on AWS.'
        ),
    )
    parser.add_argument(
        '--alias',
        dest='alias',
        type=str,
        required=True,
        help=(
            'REQUIRED. Name of the alias.'
        ),
    )
    parser.add_argument(
        '--description',
        dest='description',
        type=str,
        required=True,
        help=(
            'REQUIRED. The description of the lambda function version used for'
            'the alias.'
        ),
    )
    return parser


def exception_handler(
    error_msg: str, err: Exception, env_vars_dev: Dict, func_name: str,
) -> None:
    """Log exception and error message, and clean up the procedure.

    :param error_msg: The error message to log.
    :param err: The exception itself.
    :param env_vars_dev: Environment variables for the dev version of Lambda
        function
    :param func_name: Name of the target function.
    """
    logger.exception(err)
    logger.error(f'\033[30;41m{error_msg}\033[0m')
    if env_vars_dev:
        clean_up(env_vars_dev, func_name)
    sys.exit(1)


def update_env_variables(env_vars: Dict, func_name: str) -> Dict:
    """Update the environment variables of the Lambda.

    Only change the environment variables of the target lambda, such as the
    "ENV" field. Note that the entire environment variables must be passed, not
    just the field to be updated.

    :param env_vars: All the environment variables, including the field to be
        updated and the fields that remain the same.
    :param func_name: Name of the lambda function.
    :return: The response from calling update_function_configuration()
    :raises RuntimeError: when update_function_configuration() call fails.
    """
    try:  # config
        resp = CLIENT.update_function_configuration(
            FunctionName=func_name,
            Environment={
                'Variables': env_vars,
            },
        )
    except Exception as err:
        raise RuntimeError(
            'Update lambda function environment variable FAILED.',
        ) from err
    logger.info(resp)
    return resp


def publish_version(
    func_name: str,
    description: str,
    codesha256: str,
    revision_id: str,
) -> Dict:
    """Publish a new version of the Lambda function.

    :param func_name:   Name of the lambda function.
    :param description: A description for the new verion.
    :param codesha256:  Hashed value of the code file, obtained from the call
                        to update lambda function configuration.
    :param revision_id: Acquired from the call to update lambda function
                        configuration.
    :return: The response from calling publish_version().
    :raises RuntimeError: when publish version fails.
    """
    try:
        resp = CLIENT.publish_version(
            FunctionName=func_name,
            CodeSha256=codesha256,
            Description=description,
            RevisionId=revision_id,
        )
    except Exception as err:
        raise RuntimeError(
            'Publish new Lambda function version FAILED.',
        ) from err
    logger.info(resp)
    return resp


def get_alias(func_name: str, alias: str):
    """Get the alias of the given lambda function.

    If the given alias does not exist in the lambda function, return an empty
    dict. Otherwise, return the response of get_alias().

    :param func_name:   Name of the lambda function.
    :param alias:       Name of the alias.
    :return: Empty dict if alias doesn't exist; otherwise resp of get_alias().
    :raises RuntimeError: when get_alias() call fails.
    """
    try:
        resp_get_alias: Dict = CLIENT.get_alias(
            FunctionName=func_name, Name=alias,
        )
    except CLIENT.exceptions.ResourceNotFoundException:
        return {}
    except Exception as err:
        raise RuntimeError(
            f'Error at acquiring current aliases for {func_name}',
        ) from err
    logger.info(resp_get_alias)
    return resp_get_alias


def assign_alias(
    alias: str,
    func_name: str,
    version: str,
) -> Dict:
    """Assign an alias to the version provided.

    This function first checks whether the given alias exists. If it does not
    exist, we create a new alias. If it does, we update the current alias.

    :param alias:       Alias of the lambda function.
    :param func_name:   Name of the lambda function.
    :param version:     The lambda function version to be assigned to the
                        alias.
    :return: The response from calling create_alias() or update_alias().
    :raises RuntimeError: when update_alias() or create_alias() call fails.
    """
    resp_get_alias: Dict = get_alias(func_name, alias)
    if resp_get_alias:
        logger.info(f'Alias {alias} already exists. Update it.')
        try:
            resp_update = CLIENT.update_alias(
                FunctionName=func_name,
                Name=alias,
                FunctionVersion=version,
                Description=f'For use in the {alias} environment',
                RevisionId=resp_get_alias['RevisionId'],
            )
        except Exception as err1:
            raise RuntimeError(f'Unable to update alias: {alias}') from err1
        logger.info(resp_update)
        return resp_update
    logger.info(f'Alias {alias} does not exists. Create it.')
    try:
        resp_create = CLIENT.create_alias(
            FunctionName=func_name,
            Name=alias,
            FunctionVersion=version,
            Description=f'For use in the {alias} environment',
        )
    except Exception as err2:
        raise RuntimeError(f'Unable to create new alias: {alias}') from err2
    logger.info(resp_create)
    return resp_create


def clean_up(env_vars_dev: Dict, func_name: str) -> None:
    """Clean up the whole process of publishing a lambda version to an alias.

    The procedure requires that the lambda environment variable to be first
    updated to the non-test condition and then proceed with the publish. This
    means after the publish completes, the $LATEST lambda function remains at
    the non-test condition. This is not what we want. We want to keep the
    default $LATEST lambda version at the test condition. Therefore, we use
    this clean_up function to reset the environment variables to the test
    condition.

    :param env_vars_dev: Environment variables for the dev version of Lambda
        function
    :param func_name: Name of the lambda function.
    """
    update_env_variables(env_vars_dev, func_name)


def read_config(config_file_path: Path) -> Dict[str, Dict]:
    """Read the config JSON file and return it as a Dict.

    :param config_file_path: The path to the config file
    :type config_file_path: Path
    :return: The config as a Dict
    :rtype: Dict[str, Dict]
    """
    try:
        with open(config_file_path, 'r') as f_obj:
            func_config = json.load(f_obj)
    except FileNotFoundError as err_config:
        exception_handler(
            f'Must provide {config_file_path} file',
            err_config,
            {},
            args.func_name,
        )
    return func_config


if __name__ == '__main__':
    # Parse command line arguments
    parser: ArgumentParser = get_argument_parser()
    args = parser.parse_args()

    if args.alias == 'dev':
        logger.error('\033[30;41mUsing "dev" as alias is forbidden.\033[0m')
        sys.exit(1)

    logger.info(
        f'\033[30;43mPublishing {args.func_name} to alias prod...\033[0m',
    )

    func_config_dev = read_config(
        ROOT_DIR.joinpath(f'{args.func_folder_path}/config.dev.json'),
    )
    func_config_alias = read_config(
        ROOT_DIR.joinpath(f'{args.func_folder_path}/config.{args.alias}.json'),
    )

    resp_update_env = update_env_variables(
        func_config_alias['vars'],
        args.func_name,
    )
    resp_publish = publish_version(
        args.func_name,
        args.description,
        resp_update_env['CodeSha256'],
        resp_update_env['RevisionId'],
    )
    assign_alias(
        args.alias,
        args.func_name,
        resp_publish['Version'],
    )
    clean_up(func_config_dev['vars'], args.func_name)
    logger.info(f'\033[30;42mPublish {args.func_name} SUCCESS!\033[0m')
